### ISSUE TITLE
[BUG] ProcessColour ocassionally gets wildly too many color channels

### ISSUE THREAD

**Issue by barentine** [python-microscopy/python-microscopy#493]
**Describe the bug**
If you throw a processcolour in front of a densitymapping module, the densitymapping module doesn't see that it's been passed a color-processed tabular type. That wouldn't really bother me, but I occasionally end up with something wild. I'm not yet if that's what's causing this catastrophe, but the print statement is in densitymapping and I wouldn't have expected us to process the color twice
![image](https://user-images.githubusercontent.com/31105780/95505104-cd9a4900-097b-11eb-91dd-28a5bd1680c3.png)

recipe:
![image](https://user-images.githubusercontent.com/31105780/95503339-3a601400-0979-11eb-83a1-1daa3b8500e0.png)

relevant modules
```
- localisations.ProcessColour:
    input: z_mapped
    output: colour_mapped
- localisations.DensityMapping:
    colours:
    - chan0
    - chan1
    inputLocalizations: colour_mapped
    manualXYBounds:
    - 0.0
    - 0.0
    - 30300.0
    - 30300.0
    outputImage: hist
    pixelSize: 25.0
    renderingModule: Histogram
    xyBoundsMode: manual
```

```
'recipe': "- localisations.ProcessColour:\n    input: z_mapped\n    output: colour_mapped\n- tablefilters.FilterTable:\n    filters:\n      A:\n      - 5\n      - 20000\n      LLH:\n      - -250.0\n      - 0.0\n      error_x:\n      - 0\n      - 30\n      error_y:\n      - 0\n      - 30\n      piezoMoving:\n      - -0.5\n      - 0.5\n    inputName: Localizations\n    outputName: filtered_localizations\n- tablefilters.FilterTable:\n    filters:\n      error_x:\n      - 0\n      - 30.0\n      error_y:\n      - 0\n      - 30.0\n    inputName: filtered_localizations\n    outputName: filtered_input\n- multiview.Fold:\n    input_name: filtered_input\n    output_name: folded\n- multiview.ShiftCorrect:\n    input_name: folded\n    output_name: shift_corrected\n    shift_map_path: 'pyme-cluster://villains/CALIBRATION/2020_9_28_reg.sf'\n- multiview.FindClumps:\n    input_name: shift_corrected\n    output_name: with_clumps\n    probe_aware: true\n    radius_offset: 150.0\n- multiview.MergeClumps:\n    input_name: with_clumps\n    output_name: clumped\n- multiview.MapAstigZ:\n    astigmatism_calibration_location: 'pyme-cluster://villains/CALIBRATION/2020_9_28_psf1_astiggauss.am'\n    input_name: clumped\n    output_name: z_mapped\n    rough_knot_spacing: 100.0\n- localisations.AddPipelineDerivedVars:\n    inputEvents: ''\n    inputFitResults: FitResults\n    outputLocalizations: Localizations\n- localisations.DensityMapping:\n    colours:\n    - chan0\n    - chan1\n    inputLocalizations: colour_mapped\n    manualXYBounds:\n    - 0.0\n    - 0.0\n    - 30300.0\n    - 30300.0\n    outputImage: hist\n    pixelSize: 25.0\n    renderingModule: Histogram\n    xyBoundsMode: manual\n
```
Note that this is multiview metadata, non-ratiometric/splitter, so we _should_ have `'probe'` of `0` or `1`.

**Expected behavior**
chan 0 and chan 1 only

**Platform (please complete the following information - or copy-paste from error dialog if shown):**
 - OS: [e.g. Win10] ubuntu 14.04
 - Python version: 3.6
 - numpy version: 1.14

**Comment by barentine** [comment]
Note: removing the process colour essentially makes it one color, as the `tabular.ColourFilter` doesn't know about the `probe` key we pipe in from multiview. ProcessColour definitely needs to be there somewhere for multiview data (or change multiview.Fold to map p_chan0 etc rather than jumping straight to probe).

**Comment by barentine** [comment]
![image](https://user-images.githubusercontent.com/31105780/95523265-30034180-099c-11eb-853b-f5954b71b2a5.png)

**Comment by barentine** [comment]
Seems like a regression in `Multiview.FindClumps`, specifically the `Probe aware` bit
![image](https://user-images.githubusercontent.com/31105780/95523510-c5063a80-099c-11eb-914b-62dfdceac297.png)

**Comment by barentine** [comment]
Does not look deterministic - which would explain why the max probe is 17.6 -> 18 last time round, but recipe running on the cluster had a colourfilter with 29
![image](https://user-images.githubusercontent.com/31105780/95524014-3a263f80-099e-11eb-8a7f-6b9b9c399a50.png)

**Comment by barentine** [comment]
Well, maybe not - forgot that the initial recipe had more filters set up in the FilterTable

**Comment by barentine** [comment]
Definitely not deterministic 
![image](https://user-images.githubusercontent.com/31105780/95524646-eb79a500-099f-11eb-959c-b0b564986ff1.png)
Going to try and sort out whether it has anything to do with the order things are evaluated in (displayed/called through the shell) due to the lazy mappings involved. I guess the other thing to do would be to write a beefy unit test on find_clumps_within_channel

**Comment by barentine** [comment]
Looks like it only ever hits the 0th point.  Putting in log/raises inside of `PYME.Analysis.points.multiview.coalesce_dict_sorted` I have yet to see probe > 1 go in or come out. I don't see how 'probe' could not be evaluated by that point, so calling `np.any(inD['probe'] > 1)` shouldn't have any effect on the results, but I have yet to recreate this with those checks present.

**Comment by barentine** [comment]
Ah, that wasn't a good enough check. And issue doesn't appear to be limited to 'probe'
```
>>> pipeline['probe'].max()
1.0
>>> pipeline['z']
array([-5.74492715e+23,  5.51884294e+04,  5.38994294e+04, ...,
        4.88604294e+04,  4.83934294e+04,  4.88454294e+04])
>>> pipeline['z'].max()
55796.42942942943
>>> pipeline['z'].min()
-5.744927152950596e+23
>>> pipeline.keys()
['sigmax1', 'y_raw', 'error_sigmay2', 'error_sigmay0', 'y', 'error_sigmax2', 't', 'sigmax0', 'error_sigmay1', 'z', 'error_sigmax0', 'error_sigmax1', 'x_raw', 'tIndex', 'astigmatic_z', 'sigmax2', 'sigmay1', 'error_y', 'error_sigmay3', 'sigmay3', 'LLH', 'sigmax3', 'z_raw', 'clumpIndex', 'error_x', 'astigmatic_z_lookup_error', 'focus', 'probe', 'sigmay2', 'A', 'multiviewChannel', 'sigmay0', 'x', 'error_sigmax3']
>>> pipeline['astigmatic_z']
array([-1199.57057057,  1131.42942943,  -157.57057057, ...,
         738.42942943,   271.42942943,   723.42942943])
>>> pipeline['z']
array([-5.74492715e+23,  5.51884294e+04,  5.38994294e+04, ...,
        4.88604294e+04,  4.83934294e+04,  4.88454294e+04])
>>> pipeline['focus']
array([    0., 54057., 54057., ..., 48122., 48122., 48122.], dtype=float32)
>>> pipeline['multiviewChannel']
array([0. , 1.5, 0. , ..., 1. , 1. , 1. ], dtype=float32)
>>> pipeline['probe']
array([1.e-45, 0.e+00, 0.e+00, ..., 1.e+00, 1.e+00, 1.e+00], dtype=float32)
>>> 
```

**Comment by barentine** [comment]
For some reason can't edit my posts at the moment, but would like to add that the 'focus' key is fine after `AddPipelineDerivedVars`, and there are no zero / negative values present
```
pipeline.recipe.namespace['Localizations']['focus']
array([54057., 54057., 54057., ..., 48122., 48122., 48122.])
pipeline.recipe.namespace['Localizations']['focus'].min()
43917.0
```

**Comment by barentine** [comment]
Issue is that we only feed in clumps following PYME's cluster numbering of zero being unclustered and 1 being the first cluster. Passing this into `PYME.Analysis.points.DeClump.deClump.c`, `aggregateMean` is therefore problematic as the output is only ever written with `outVar[currentClump]`.

`currentClump` starts at -1 rather than 0, and is only ever set to values in `clumpIndex`, i.e. starting at 1, so `outVar[0]` is left at whatever it was initialized at in `outVar = (float*)PyArray_DATA(outVarA);`

In other words, `aggregateMean` is busted. It assumes the data has been sorted by `clumpIndex` before being passed in, which means we actually want to start `currentClump` at `clumpIDs[0]` and write into `outVar` with an index of `currentClump - clumpID_0` or so. Doing so will then actually output an array of the right length after aggregation. What the output of aggregateMean should be for a point belonging to group 0 is a bit beyond me - I think it's safer if these aggregate methods are only passed data where unclumped points have their own clumpIndex. Taking the `currentClump - clumpID_0` approach should let us work with data following either convention (starting at 1 and unclustered points get their own clumpID, and starting at 0 and unclustered points get their own clumpID).

**Comment by barentine** [comment]
Note that #511, unlike #501, leaves us with some issues

```
Traceback (most recent call last):
  File "/home/smeagol/code/python-microscopy/PYME/ui/progress.py", line 107, in func
    return fcn(*args, **kwargs)
  File "/home/smeagol/code/pyme-omero/pyme_omero/visgui_modules/omero_loader.py", line 117, in OnSaveSnapshot
    metadata=mdh)
  File "/home/smeagol/code/python-microscopy/PYME/IO/tabular.py", line 147, in to_hdf
    f.appendToTable(tablename, self.to_recarray(keys))
  File "/home/smeagol/code/python-microscopy/PYME/IO/tabular.py", line 107, in to_recarray
    columns = [self.__getitem__(k) for k in keys]
  File "/home/smeagol/code/python-microscopy/PYME/IO/tabular.py", line 107, in <listcomp>
    columns = [self.__getitem__(k) for k in keys]
  File "/home/smeagol/code/python-microscopy/PYME/IO/tabular.py", line 879, in __getitem__
    res = np.array(self.resultsSource[key])[self.Index]
IndexError: boolean index did not match indexed array along dimension 0; dimension is 1294 but corresponding boolean dimension is 1293
```


```
for k in pipeline.keys():
    print('%s, len: %d' % (k, len(pipeline[k])))
    
sigmay1, len: 1293
multiviewChannel, len: 1293
sigmay2, len: 1293
p_chan1, len: 1293
p_chan0, len: 1293
Traceback (most recent call last):
  File "<input>", line 2, in <module>
  File "/home/smeagol/code/python-microscopy/PYME/LMVis/pipeline.py", line 349, in __getitem__
    return self.output[keys]
  File "/home/smeagol/code/python-microscopy/PYME/IO/tabular.py", line 1066, in __getitem__
    return self.get_channel_column(self.currentColour, keys)
  File "/home/smeagol/code/python-microscopy/PYME/IO/tabular.py", line 1073, in get_channel_column
    return self.resultsSource[keys]
  File "/home/smeagol/code/python-microscopy/PYME/IO/tabular.py", line 723, in __getitem__
    return self.resultsSource[key][self.Index][sl]
  File "/home/smeagol/code/python-microscopy/PYME/IO/tabular.py", line 923, in __getitem__
    return self.resultsSource[keys]
  File "/home/smeagol/code/python-microscopy/PYME/IO/tabular.py", line 879, in __getitem__
    res = np.array(self.resultsSource[key])[self.Index]
IndexError: boolean index did not match indexed array along dimension 0; dimension is 1294 but corresponding boolean dimension is 1293

```

Have yet to sort out why.

**Comment by barentine** [comment]
ah, it's on the weights, e.g. `error_x` etc. Need to errVec[non_trivial] too. Will sub a PR with that, docstring -> numpy, etc.

**Comment by barentine** [comment]
Note that with or without that errVec[non_trivial] fix, the current head breaks on the tests just added
```
$ pytest tests/PYME/Analysis/points/test_clumps.py 
============================================================== test session starts ==============================================================
platform linux -- Python 3.6.9, pytest-5.3.5, py-1.9.0, pluggy-0.13.1
rootdir: /home/smeagol/code/python-microscopy
collected 4 items                                                                                                                               

tests/PYME/Analysis/points/test_clumps.py FF..                                                                                            [100%]

=================================================================== FAILURES ====================================================================
_________________________________________________________________ test_coalesce _________________________________________________________________

    def test_coalesce():
        from PYME.Analysis.points import multiview
        import numpy as np
        assigned = np.array([0,0,1,1,2,3,4,5,5,5,6], 'i')
        x = np.random.rand(len(assigned))
        x_out = multiview.coalesce_dict_sorted({'x':x}, assigned, ['x',], {})['x']
    
        for j in np.unique(assigned):
>           assert (x_out[j] == np.mean(x[assigned==j]))
E           AssertionError: assert 0.50802445 == 0.5080244395505363
E            +  where 0.5080244395505363 = <function mean at 0x7fb940b33b70>(array([0.66698042, 0.34906846]))
E            +    where <function mean at 0x7fb940b33b70> = <module 'numpy' from '/home/smeagol/miniconda3/lib/python3.6/site-packages/numpy/__init__.py'>.mean

tests/PYME/Analysis/points/test_clumps.py:9: AssertionError
___________________________________________________________ test_coalesce_incomplete ____________________________________________________________

    def test_coalesce_incomplete():
        from PYME.Analysis.points import multiview
        import numpy as np
        assigned = np.array([1, 1, 2, 3, 5, 5, 5, 6, 9,9], 'i')
        x = np.random.rand(len(assigned))
        x_out = multiview.coalesce_dict_sorted({'x': x}, assigned, ['x', ], {})['x']
    
        for j in np.unique(assigned):
>           assert (x_out[j] == np.mean(x[assigned == j]))
E           AssertionError: assert 0.39201748 == 0.39201746119559955
E            +  where 0.39201746119559955 = <function mean at 0x7fb940b33b70>(array([0.46966945, 0.31436547]))
E            +    where <function mean at 0x7fb940b33b70> = <module 'numpy' from '/home/smeagol/miniconda3/lib/python3.6/site-packages/numpy/__init__.py'>.mean

tests/PYME/Analysis/points/test_clumps.py:22: AssertionError
```

**Comment by barentine** [comment]
Note also that the declump init in #512 does not fix the errors we're hitting on the tests I just mentioned, nor the fail in the test added in #514

**Comment by barentine** [comment]
Fixed with #511 and #514
